services/src/main/java/org/keycloak/protocol/oidc/grants/device/endpoints/DeviceEndpoint.java

```diff
@@ -0,0 +1,488 @@
+            event.error(Errors.NOT_ALLOWED);
+            throw new CorsErrorResponseException(cors, OAuthErrorException.INVALID_GRANT,
+                "Client not allowed for OAuth 2.0 Device Authorization Grant", Response.Status.BAD_REQUEST);
+        }
+
+        int expiresIn = realm.getOAuth2DeviceConfig().getLifespan(client);
+        int interval = realm.getOAuth2DeviceConfig().getPoolingInterval(client);
+
+        OAuth2DeviceCodeModel deviceCode = OAuth2DeviceCodeModel.create(realm, client,
+            Base64Url.encode(KeycloakModelUtils.generateSecret()), request.getScope(), request.getNonce(), expiresIn, interval,
+            request.getAdditionalReqParams());
+        OAuth2DeviceUserCodeProvider userCodeProvider = session.getProvider(OAuth2DeviceUserCodeProvider.class);
+        String secret = userCodeProvider.generate();
+        OAuth2DeviceUserCodeModel userCode = new OAuth2DeviceUserCodeModel(realm, deviceCode.getDeviceCode(), secret);
+
+        // To inform "expired_token" to the client, the lifespan of the cache provider is longer than device code
+        int lifespanSeconds = expiresIn + interval + 10;
+
+        OAuth2DeviceTokenStoreProvider store = session.getProvider(OAuth2DeviceTokenStoreProvider.class);
+        store.put(deviceCode, userCode, lifespanSeconds);
+
+        try {
+            String deviceUrl = DeviceGrantType.oauth2DeviceVerificationUrl(session.getContext().getUri()).build(realm.getName())
+                .toString();
+
+            OAuth2DeviceAuthorizationResponse response = new OAuth2DeviceAuthorizationResponse();
+            response.setDeviceCode(deviceCode.getDeviceCode());
+            response.setUserCode(userCodeProvider.display(secret));
+            response.setExpiresIn(expiresIn);
+            response.setInterval(interval);
+            response.setVerificationUri(deviceUrl);
+            response.setVerificationUriComplete(deviceUrl + "?user_code=" + response.getUserCode());
+
+            return Response.ok(JsonSerialization.writeValueAsBytes(response)).type(MediaType.APPLICATION_JSON_TYPE).build();
+        } catch (Exception e) {
+            throw new RuntimeException("Error creating OAuth 2.0 Device Authorization Response.", e);
+        }
+    }
+
+    /**
+     * This endpoint is used by end-users to start the flow to authorize a device.
+     *
+     * @param userCode the user code to authorize
+     * @return
+     */
+    @GET
+    public Response verifyUserCode(@QueryParam("user_code") String userCode) {
+        event.event(EventType.OAUTH2_DEVICE_VERIFY_USER_CODE);
+
+        checkSsl();
+        checkRealm();
+
+        // So back button doesn't work
+        CacheControlUtil.noBackButtonCacheControlHeader();
+
+        AuthenticationSessionModel authenticationSession = null;
+
+        // try to recover the authentication session in case the user is pushing the code
+        if (HttpMethod.POST.equalsIgnoreCase(httpRequest.getHttpMethod())) {
```

Comment:
Instead of checking HttpMethod here, it will be slightly better if there is private method introduced like `verifyUserCode(String userCode, boolean isActionRequest)` .  When this private method is called from the GET method, it will use `isActionRequest` as false. When there is a call from the verifyUserCode POST method, it will use true. This will also allow to avoid duplicating creation of authenticationSession as I described below.

Reply:
I fixed.